### ISSUE TITLE
N3: oracle-B variants + a representation-rank instrument — why does FAITHFUL feedback hurt?

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#183]
The sharpest open mechanism question in the dfa-vs-bp programme, and the only one with no data behind it. Context: `projects/dfa-vs-bp/docs/synthesis.md` §3.2 and §7.

## The finding that needs explaining

At `ae_shak_a65` — a fixture with **no planted noise at all** — the oracle `B` routes `p_energy = 0.99999` of the error energy (it is literally the identity on the active classes) and performs **exactly like never training the body**:

| arm, n=8, own LR optimum | bpb | error energy routed |
|---|---|---|
| frozen (body never trained) | 4.132 ± .038 | — |
| **oracle `B`** | **4.125 ± .058** | **0.99999** |
| random `B`, rank 65 | **3.884** ± .055 | 0.016 |

A random projection carrying 1.6% of the error beats perfect routing by 0.240 bpb (t = +8.5), both at interior optima across 3.5 decades of LR. Rank is not the explanation (lifting the oracle to full rank changes nothing; random `B` is indifferent to rank). Same pattern at the noisy fixture: the oracle delivers precisely the learnable component and loses by 0.68 bpb.

**So DFA is not approximate backpropagation, and making the feedback more faithful makes it worse.** We can state that; we cannot explain it.

### (a) Three oracle variants, on the `review/d1b-oracle-complement` path (#176)

The oracle machinery exists; these decompose *what* about faithfulness hurts. Each is a different `P` under the existing `GTX_DFA_ADAPT=oracle`:

- **row-permuted `P`** — the same rows, shuffled. Right magnitudes and right sparsity, wrong direction. Separates "the error's magnitude structure" from "the error's direction".
- **sign-only oracle** — `sign(P)`, so the routing is preserved but the magnitudes are flattened.
- **per-step random sign flip** — the oracle with a fresh random sign per row per step, i.e. faithful direction decorrelated *across time*.

If the decorrelated variants recover, the harm is that faithful feedback is *consistent over training*, not that it is faithful.

### (b) A representation-rank instrument on the BODY's activations

This is the measurement the hypothesis needs and the lane does not have. E1 already built exactly this shape for the **error covariance** — `GTX_INSTRUMENT` computes a Gram matrix, trace, `lambda_max`, participation ratio and stable rank over `C_E`, with a sampling-ceiling detector. What N3 needs is the same statistics over the **body's hidden activations** (per block, or at the final tap), not over the error.

Reusing that machinery rather than writing a second one would be ideal — same ceiling caveat, same provenance style, same opt-in byte-null discipline.

**Hypothesis to pre-register:** faithful feedback drives the body to co-adapt with the BP-trained head and **collapses representational diversity**; near-orthogonal feedback acts as a perturbation that preserves it. Prediction: activation effective rank falls fastest in the oracle arm, is preserved in the random arm, and tracks bpb across all the variants above.

If that is wrong, it is wrong in an interesting way — the instrument decides it either direction, which is why it is worth building before more arms are run.

## Why it matters beyond curiosity

Three architecture attempts have now been priced by this question without answering it: F9d, Design A (N1b/N1c) and Design B (B0b) all turn on *where DFA capacity is worth putting*, and the rule we extracted — "only where nothing else can do that job" — is descriptive. Knowing why faithful feedback hurts would make it predictive, which is the difference between a heuristic and a design principle.

Related: #176 (the oracle path), #182 (cost), #177, #178.

**Comment by OriPekelman** [comment]
**(b) is shipped. (a) is blocked on #176 — read that part first, it changes what (a) should look like.**

`main` **9ebf697**, battery `ALL GATES PASS (57 legs)`.

## (b) The instrument

`GTX_ACTRANK=1` / `GTX_ACTRANK_N`:

```
actrank: n= d= batches= trace= lambda_max= participation_ratio= stable_rank= ceiling= capped=
```

Reuses E1 rather than duplicating it, as you asked: `bc_gram!` / `bc_trace` / `bc_fro2` / `bc_lambda_max` already take the vector dimension as a parameter, so passing `D_MODEL` instead of `N_CLASSES` is the entire change — and the two lines stay comparable *because* they are computed by the same code. Sampled from the same val pass as `bcond`, so both instruments see identical batches.

Hook is `t_h` at the bytelm head, taken after the detach (detach changes graph edges, not values, so these are the body's activations either way).

**Byte-null in both senses**: no line when off, and the bpb is **bit-identical** with it on — `4.432748299174562` either way. That property is what lets it be left enabled, and it is the gate's load-bearing assertion. An instrument that perturbs its own cell would mean every arm it compares was measured on a different run than the one that scored it.

**The ceiling bites harder here than in E1, and asymmetrically.** `rank <= min(n, d_model)`, and `d_model` is usually far below the vocabulary, so a participation ratio near `d_model` is bounded by the model's *width* rather than measured. Your hypothesis is about **collapse**, so a ceiling reads as **high** rank — the direction that looks like the hypothesis being refuted. Both bounds and a `capped=` flag are reported, and the gate fails if the cell is capped.

## (a) is blocked, and not on a technicality

All three variants are `P` under `GTX_DFA_ADAPT=oracle`. On `main` that flag accepts **`none|oja`** — there is no `oracle` token. The oracle path is on `review/d1b-oracle-complement` (#176), still unmerged.

So (a) cannot be built, and — more importantly — **the pre-registered prediction is untestable**: it names the oracle arm as the one where rank should fall fastest, and that is the one arm that does not exist on `main`.

## First measurement, and a confound worth designing around

2 blocks, `d_model=64`, `n=256`, all `capped=ok`:

| arm | participation_ratio | stable_rank | bpb |
|---|---|---|---|
| chain | **1.445** | 1.206 | **4.433** |
| frozen | 2.512 | 1.619 | 5.422 |
| dfa (random B) | **2.610** | 1.761 | **5.505** |

**Rank moves WITH bpb, not against it.** The best arm has the lowest rank. Your hypothesis has collapse accompanying harm; here it accompanies the best result available.

Before that is read as evidence either way — **`frozen` sits at 2.51 with an untrained body**, i.e. near random-init diversity. A simpler story fits all three rows: *training collapses rank, and more effective training collapses it more.* That would make rank a proxy for **how much the body was shaped**, not a cause of quality.

If that confound is real it also threatens the oracle prediction directly: the oracle arm performs like never training the body, so it would sit at the *high*-rank end for the same trivial reason `frozen` does — and a rank/bpb correlation across your variants could be mostly "how much did the body move".

**Suggestion for (a):** include a control that separates *amount* of training from *faithfulness* of feedback — e.g. random `B` at matched update magnitude, or the same arm sampled at matched training loss rather than matched steps. Otherwise the row-permuted / sign-only / sign-flip variants may all differ in rank for a reason that has nothing to do with faithfulness.

These are also **not your arms and not your fixture** — you measured on `ae_shak_a65` where random `B` beat frozen; here `dfa` is marginally *worse* than frozen. I would not read across, and I am not claiming a result about N3. The instrument is delivered and gated; the question is still open.

Ready to build (a) as soon as #176 lands.

**Comment by OriPekelman** [comment]
**(a) is unblocked.** `main` **9f7aace** — #176 merged, and the missing piece is now committed.

The three variants were blocked on two things, not one:

1. `GTX_DFA_ADAPT=oracle` absent from `main` — **fixed**, #176 merged as `b7dbb76`.
2. no pooling map pack to run it against — **fixed**, `data/oracle_map_c256_b65` (1.4 KB) plus `prep/build_oracle_map.rb`.

The oracle now runs on `main` and its behaviour is gated, so the variants have a path that is asserted rather than merely present.

## But read this before building them

**The fixture is a MECHANICS fixture, not a faithful oracle.** `modulo` pooling routes `p_energy = 0.243`, against D1's `0.99999`. It is arbitrary by design — a gate wants predictable structure, not a pooling that encodes a hypothesis about which codes belong together.

So the variants' **mechanics** can be built and gated today (each transforms `P` after `init_p_from_map!`, a single clean hook), but their **interpretation** still needs a map that routes ~1.0 of the error energy. Row-permuted vs sign-only vs sign-flip only decomposes *faithfulness* if the thing being perturbed is faithful to begin with. That map is an inflation-fixture design decision and belongs to you.

## First oracle numbers, with the (b) instrument on the same runs

2 blocks, `ae_shakespeare`, `GTX_LDFA_EVERY=10`:

| arm | bpb | participation_ratio |
|---|---|---|
| chain | 4.433 | 1.445 |
| frozen | 5.422 | 2.512 |
| dfa (random B) | 5.505 | 2.610 |
| oracle (modulo) | 5.686 | **2.483** |

**The oracle lands at the high-rank end**, near `frozen` (2.512) and well above `chain` (1.445) — exactly where the confound I raised predicts it would, for a reason that has nothing to do with faithfulness.

Recall the alternative story: *training collapses rank, and more effective training collapses it more.* `frozen` sits high because its body is untrained; the oracle sits high because it performs like never training the body. On that reading, rank tracks **how much the body moved**, and a rank/bpb correlation across your variants would be largely mechanical.

This is a weak instance — the map is not faithful, so it is not the oracle your hypothesis names — but it is the direction the confound predicts, and it is now measurable rather than hypothetical.

**Suggestion for (a), unchanged but now concrete:** include an arm that separates *amount of training* from *faithfulness of feedback* — random `B` at matched update magnitude, or arms compared at matched training loss rather than matched steps. Otherwise row-permuted / sign-only / sign-flip may all differ in rank because they differ in how much the body moved, and the hypothesis would look confirmed either way.

Ready to build the three variants on that basis whenever you want them — mechanics gated now, interpretation pending a faithful map.

**Comment by OriPekelman** [comment]
**`GTX_ACTRANK` ran, and it refuted the hypothesis it was built for.** Which is the best possible outcome for a pre-registered prediction, and worth reporting promptly since it changes what part (a) is for.

Gate first: with the instrument ON, the oracle cell reproduces D1b's `4.087987132927531` bit-for-bit. It does not move the number it sits beside. `capped=ok` asserted on every cell.

`ae_shak_a65` (no planted noise), each arm at its own D1b-bracketed optimum, n=5, ceiling d=128:

| arm | bpb | participation | stable rank |
|---|---|---|---|
| bp-body | 3.314 ± .031 | 5.22 ± .31 | 2.41 ± .08 |
| frozen-body | 4.130 ± .048 | 3.14 ± .66 | 1.82 ± .21 |
| dfa-wide | 3.967 ± .089 | 9.53 ± 1.07 | 4.54 ± .72 |
| **dfa-fix-65** (random P) | **3.919 ± .032** | 11.09 ± .66 | **5.33 ± .54** |
| **dfa-orc-plain** (oracle P) | **4.127 ± .064** | 11.42 ± 1.09 | **5.33 ± .18** |

## The hypothesis was wrong

Pre-registered: *faithful feedback collapses representational diversity; near-orthogonal feedback preserves it; rank tracks bpb.*

Measured, oracle vs random at matched rank 65: **bpb +0.208 (t = +6.51)**, **participation +0.33 (t = +0.58, n.s.)**, **stable rank +0.00 (t = 0.00)**. The two arms have statistically identical representational diversity and are 0.208 bpb apart. The oracle does not collapse anything.

Across all five arms the correlation between participation ratio and bpb is **r = +0.29** — the *wrong sign*, and driven by BP, which reaches the best loss (3.314) on a **lower**-diversity representation (5.22) than any DFA arm. Diversity is not the explanatory variable, in either direction.

## What the instrument did find, which is new

**DFA representations are roughly twice as diverse as BP's** — participation 9.5–11.4 against BP's 5.22, stable rank 4.5–5.3 against 2.41 — while being worse at the task. And frozen sits lowest (3.14). So DFA does spread the representation out relative to BP, consistent with the perturbation/regulariser reading elsewhere in the programme; it just is not what separates a good DFA arm from a bad one.

## What this does to part (a)

It **raises** its value. The rank story was the leading candidate explanation and it is now eliminated, so the oracle variants stop being a decomposition of a known mechanism and become the primary probe:

- **row-permuted `P`** — right magnitudes and sparsity, wrong direction
- **sign-only oracle** — routing preserved, magnitudes flattened
- **per-step random sign flip** — faithful direction, decorrelated across time

That last one is now the most interesting, because the surviving candidate after this result is *consistency over training* rather than anything about the representation: the random `B` is a fixed arbitrary map, the oracle is a fixed faithful one, and they differ in what they are correlated with, not in how much diversity they leave behind.

No rush on my side — the finding stands on its own and is being written up as a refuted hypothesis either way. Raw: `gx10:/srv/data/scratch/rev0828/n3/`.
